### ISSUE TITLE
Inference failure of TensorRT 10.0.x.x when running my internal model on GPU(T4, A100)

### ISSUE THREAD

**Issue by kimdwkimdw** [NVIDIA/TensorRT#4066]
## Description

After updating to TensorRT 10.0.1.6, we expected the previously reported issue to be resolved. Unfortunately, not only does the issue persist, but the model’s outputs have deteriorated even further. Specifically, all output values are now nan, making it impossible to use our models. This issue affects both fp16 and fp32 precision settings, rendering the model completely non-functional.

https://github.com/NVIDIA/TensorRT/issues/3292

## Environment

TensorRT Version: All version of 10.0.x.x. NGC Container 24.05~24.07.

NVIDIA GPU: T4, A100

NVIDIA Driver Version: 550.90.07

CUDA Version: 12.4

CUDNN Version: x

Operating System:

Container (if so, version): NGC Container from 23.03 and 24.07.
https://catalog.ngc.nvidia.com/orgs/nvidia/containers/tensorrt

Operating System:

Python Version (if applicable):

Tensorflow Version (if applicable):

PyTorch Version (if applicable):

Baremetal or Container (if so, version):

## Relevant Files

**Model link**:

## Steps To Reproduce

**Commands or scripts**:

**Have you tried [the latest release](https://developer.nvidia.com/tensorrt)?**:

**Can this model run on other frameworks?** For example run ONNX model with ONNXRuntime (`polygraphy run <model.onnx> --onnxrt`):

**Comment by lix19937** [comment]
Can you provide the full log with `trtexec --verbose` ?

**Comment by kimdwkimdw** [comment]
@lix19937 Can I send this to you via email? I prefer not to expose my model publicly. The --verbose option also reveals too much information.

**Comment by lix19937** [comment]
@kimdwkimdw  ok, my email is jevenlee2016@foxmail.com

**Comment by kimdwkimdw** [comment]
@lix19937 i've sent it with gzipped

**Comment by lix19937** [comment]
@kimdwkimdw  I didn't receive your e-mail.

**Comment by kimdwkimdw** [comment]
> @kimdwkimdw  I didn't receive your e-mail.   

Please check your spam mail inbox.

Or, search with "arthur at rtzr.ai"

I've sent mail via Gmail.

**Comment by lix19937** [comment]
Can you upload the goole drive ?

**Comment by kimdwkimdw** [comment]
> Can you upload the goole drive ?

OK I upload log file to gdrive and shared with your email

**Comment by kimdwkimdw** [comment]
Any updates?

**Comment by kimdwkimdw** [comment]
@lix19937 let me know your google address for Google drive. `[jevenlee2016@foxmail.com](mailto:jevenlee2016@foxmail.com)` seems doesn't work.

**Comment by kimdwkimdw** [comment]
This is same kind of issue from https://github.com/NVIDIA/TensorRT/issues/3292 

TensorRT 10.x have significant errors.

cc. @zerollzeng @ttyio

**Comment by lix19937** [comment]
> [@lix19937](https://github.com/lix19937) let me know your google address for Google drive. Your email address `jevenlee2016@foxmail.com` seems doesn't work.

sent log to email  hblijinwen@126.com

**Comment by kimdwkimdw** [comment]
@lix19937 I've sent it to hblijinwen@126.com

**Comment by lix19937** [comment]
@kimdwkimdw    

> but the model’s outputs have deteriorated even further. Specifically, all output values are now nan, making it impossible to use our models.   

From your logs, it has no valid errors or warnings. Maybe you can use polygraphy, like follow 
```shell  
polygraphy run model.onnx --trt  --onnxrt --input-shapes source:[2,160000] wav_lens:[2,1]
``` 
to check which layer begin to arise the big **nan/diff**, check whether a BN after conv, etc. Also you can check the weights max-min range.    

Another hand, you can try to use latest version.

**Comment by kimdwkimdw** [comment]
@lix19937 

Thank you for the suggestion, but I have already tried using `polygraphy` along with other methods. My question goes back to the root of the issue: why do nan values in the relative difference output from polygraphy start appearing when using TensorRT version 10.x?

I did not encounter this issue with TensorRT 8.5.3, including versions like 23.02 and 23.03, where there were no nan values. However, starting from TensorRT 8.6.1.6, the errors have become more pronounced, and with all versions of TensorRT 10 (e.g., 10.3.0.26, 10.2.0, 10.1.0), the model’s errors seem to overflow dramatically.

**Comment by lix19937** [comment]
In my opinion, form 8.6, tensorrt add more feature, like builder optimization level(the default optimization level is 3. Valid values include integers from 0 to the maximum optimization level 5), and import llm layers fusion(like mha, ln), for normalization layers, particularly if deploying with mixed precision, target the latest ONNX opset that contains the corresponding function ops, for example: opset 17 for LayerNormalization or opset 18 GroupNormalization. Numerical accuracy using function ops is superior to corresponding implementation with primitive ops for normalization layers. And move cuda-x lib depends.    

You can try to add `--builderOptimizationLevel=5  --noTF32`  and adjust the size of `--memPoolSize` . @kimdwkimdw

**Comment by kimdwkimdw** [comment]
I have already tried the suggestions, including setting --builderOptimizationLevel=5 and --noTF32, and updating the ONNX opset from 15 to 17. Currently, proper export for opset 18 is not supported in PyTorch (Ref: [PyTorch ONNX Export](https://pytorch.org/docs/2.3/onnx_dynamo.html#torch.onnx.ExportOptions)).

Despite these adjustments, including experimenting with different precision combinations (fp16, bf16, tf32, fp32 with --noTF32), the output from polygraphy continues to show significant degradation as follows:

```
[I]         Error Metrics: x
[I]             Minimum Required Tolerance: elemwise error | [abs=11.761] OR [rel=1393.1] (requirements may be lower if both abs/rel tolerances are set)
[I]             Absolute Difference | Stats: mean=1.6023, std-dev=1.614, var=2.605, median=1.0086, min=0 at (1, 234, 1208), max=11.761 at (0, 14, 619), avg-magnitude=1.6023, p90=4.0612, p95=4.0612, p99=6.56
[I]                 ---- Histogram ----
                    Bin Range    |  Num Elems | Visualization
                    (0   , 1.18) |    8723143 | ########################################
                    (1.18, 2.35) |    3338792 | ###############
                    (2.35, 3.53) |    1736355 | #######
                    (3.53, 4.7 ) |    1026219 | ####
                    (4.7 , 5.88) |     718419 | ###
                    (5.88, 7.06) |     308960 | #
                    (7.06, 8.23) |      65386 | 
                    (8.23, 9.41) |      16262 | 
                    (9.41, 10.6) |       2064 | 
                    (10.6, 11.8) |        400 | 
[I]             Relative Difference | Stats: mean=0.1035, std-dev=0.84023, var=0.70599, median=0.068093, min=0 at (1, 234, 1208), max=1393.1 at (0, 107, 0), avg-magnitude=0.1035, p90=0.24082, p95=0.24082, p99=0.40837
[I]                 ---- Histogram ----
                    Bin Range            |  Num Elems | Visualization
                    (0       , 139     ) |   15935951 | ########################################
                    (139     , 279     ) |         36 | 
                    (279     , 418     ) |          3 | 
                    (418     , 557     ) |          4 | 
                    (557     , 697     ) |          2 | 
                    (697     , 836     ) |          0 | 
                    (836     , 975     ) |          1 | 
                    (975     , 1.11e+03) |          0 | 
                    (1.11e+03, 1.25e+03) |          0 | 
                    (1.25e+03, 1.39e+03) |          3 | 
```

This indicates a severe accuracy issue. Features that were functional in TensorRT 8.5 no longer work correctly in the later versions.

@lix19937

**Comment by lix19937** [comment]
> I have already tried the suggestions, including setting `--builderOptimizationLevel=5` and `--noTF32`, and updating the ONNX opset from 15 to 17. Currently, proper export for opset 18 is not supported in PyTorch (Ref: [PyTorch ONNX Export](https://pytorch.org/docs/2.3/onnx_dynamo.html#torch.onnx.ExportOptions)).
> 
> Despite these adjustments, including experimenting with different precision combinations (fp16, bf16, tf32, fp32 with --noTF32), the output from polygraphy continues to show significant degradation as follows:
> 
> ```
> [I]         Error Metrics: x
> [I]             Minimum Required Tolerance: elemwise error | [abs=11.761] OR [rel=1393.1] (requirements may be lower if both abs/rel tolerances are set)
> [I]             Absolute Difference | Stats: mean=1.6023, std-dev=1.614, var=2.605, median=1.0086, min=0 at (1, 234, 1208), max=11.761 at (0, 14, 619), avg-magnitude=1.6023, p90=4.0612, p95=4.0612, p99=6.56
> [I]                 ---- Histogram ----
>                     Bin Range    |  Num Elems | Visualization
>                     (0   , 1.18) |    8723143 | ########################################
>                     (1.18, 2.35) |    3338792 | ###############
>                     (2.35, 3.53) |    1736355 | #######
>                     (3.53, 4.7 ) |    1026219 | ####
>                     (4.7 , 5.88) |     718419 | ###
>                     (5.88, 7.06) |     308960 | #
>                     (7.06, 8.23) |      65386 | 
>                     (8.23, 9.41) |      16262 | 
>                     (9.41, 10.6) |       2064 | 
>                     (10.6, 11.8) |        400 | 
> [I]             Relative Difference | Stats: mean=0.1035, std-dev=0.84023, var=0.70599, median=0.068093, min=0 at (1, 234, 1208), max=1393.1 at (0, 107, 0), avg-magnitude=0.1035, p90=0.24082, p95=0.24082, p99=0.40837
> [I]                 ---- Histogram ----
>                     Bin Range            |  Num Elems | Visualization
>                     (0       , 139     ) |   15935951 | ########################################
>                     (139     , 279     ) |         36 | 
>                     (279     , 418     ) |          3 | 
>                     (418     , 557     ) |          4 | 
>                     (557     , 697     ) |          2 | 
>                     (697     , 836     ) |          0 | 
>                     (836     , 975     ) |          1 | 
>                     (975     , 1.11e+03) |          0 | 
>                     (1.11e+03, 1.25e+03) |          0 | 
>                     (1.25e+03, 1.39e+03) |          3 | 
> ```
> 
> This indicates a severe accuracy issue. Features that were functional in TensorRT 8.5 no longer work correctly in the later versions.
> 
> [@lix19937](https://github.com/lix19937)

can you provide the result of polygraphy compare trt with onnxrt in trt8.5 ?

**Comment by moraxu** [comment]
@kimdwkimdw can you send the model to mguzek@nvidia.com, I can instance an internal bug for this

**Comment by kimdwkimdw** [comment]
@moraxu I've shared google drive link with `mguzek@nvidia.com`. I've shared it with `zeroz@nvidia.com` last year.

**Comment by yuanyao-nv** [comment]
@kimdwkimdw Can you please also try the latest TRT 10.5? There are some known accuracy issues that have been fixed in 10.5.

**Comment by kimdwkimdw** [comment]
@yuanyao-nv 

Currently, all of our tests are being conducted using the following TensorRT container, but it appears that version 10.5 is not yet included in the 24.09 release:
TensorRT 10.4.0.26 in NGC 24.09 - https://catalog.ngc.nvidia.com/orgs/nvidia/containers/tensorrt/tags 

Once an update is available, I will test with version 10.5.

**Comment by kimdwkimdw** [comment]
@yuanyao-nv "NGC 24.10 with TensorRT 10.5.0.18" updated recently. I've tested all of tests with TRT 10.5.

Nothing improved.
